pilot/pkg/model/authorization.go

```diff
@@ -30,162 +29,168 @@ type RolesAndBindings struct {
 		return
 	}
-	if policy.NamespaceToPolicies == nil {
-		policy.NamespaceToPolicies = map[string]*RolesAndBindings{}
+	if _, ok := role.Spec.(*rbacproto.ServiceRole); !ok {
+		return
 	}
-	if policy.NamespaceToPolicies[role.Namespace] == nil {
-		policy.NamespaceToPolicies[role.Namespace] = &RolesAndBindings{
-			Roles:              []Config{},
-			RoleNameToBindings: map[string][]*rbacproto.ServiceRoleBinding{},
+
+	if policy.namespaceToV1Policies[role.Namespace] == nil {
+		policy.namespaceToV1Policies[role.Namespace] = &RolesAndBindings{
+			Bindings: map[string][]*rbacproto.ServiceRoleBinding{},
 		}
 	}
-	rolesAndBindings := policy.NamespaceToPolicies[role.Namespace]
+	rolesAndBindings := policy.namespaceToV1Policies[role.Namespace]
 	rolesAndBindings.Roles = append(rolesAndBindings.Roles, *role)
-
-	// Initialize NameToServiceRoles for Authz v2.
-	if policy.NamespaceToAuthorizationConfigV2 == nil {
-		policy.NamespaceToAuthorizationConfigV2 = map[string]*AuthorizationConfigV2{}
-	}
 }
 
 func (policy *AuthorizationPolicies) addServiceRoleBinding(binding *Config) {
-	if binding == nil || binding.Spec.(*rbacproto.ServiceRoleBinding) == nil {
+	if binding == nil || policy == nil {
 		return
 	}
-	name := binding.Spec.(*rbacproto.ServiceRoleBinding).RoleRef.Name
+	spec, ok := binding.Spec.(*rbacproto.ServiceRoleBinding)
+	if !ok {
+		return
+	}
+
+	name := spec.RoleRef.Name
 	if name == "" {
 		rbacLog.Errorf("ignored invalid binding %s in %s with empty RoleRef.Name",
 			binding.Name, binding.Namespace)
 		return
 	}
-	if policy.NamespaceToPolicies == nil {
-		policy.NamespaceToPolicies = map[string]*RolesAndBindings{}
-	}
-	if policy.NamespaceToPolicies[binding.Namespace] == nil {
-		policy.NamespaceToPolicies[binding.Namespace] = &RolesAndBindings{
-			Roles:              []Config{},
-			RoleNameToBindings: map[string][]*rbacproto.ServiceRoleBinding{},
+
+	if policy.namespaceToV1Policies[binding.Namespace] == nil {
+		policy.namespaceToV1Policies[binding.Namespace] = &RolesAndBindings{
+			Bindings: map[string][]*rbacproto.ServiceRoleBinding{},
 		}
 	}
-	rolesAndBindings := policy.NamespaceToPolicies[binding.Namespace]
-	if rolesAndBindings.RoleNameToBindings[name] == nil {
-		rolesAndBindings.RoleNameToBindings[name] = []*rbacproto.ServiceRoleBinding{}
-	}
-	rolesAndBindings.RoleNameToBindings[name] = append(
-		rolesAndBindings.RoleNameToBindings[name], binding.Spec.(*rbacproto.ServiceRoleBinding))
+	rolesAndBindings := policy.namespaceToV1Policies[binding.Namespace]
+	rolesAndBindings.Bindings[name] = append(
+		rolesAndBindings.Bindings[name], binding.Spec.(*rbacproto.ServiceRoleBinding))
 }
 
-// AddConfig adds a config of type ServiceRole, ServiceRoleBinding to AuthorizationPolicies.
-func (policy *AuthorizationPolicies) AddConfig(cfgs ...*Config) {
+func (policy *AuthorizationPolicies) addConfig(cfgs []Config) {
 	for _, cfg := range cfgs {
-		if cfg == nil {
-			continue
-		}
 		switch cfg.Spec.(type) {
```

Comment:
Can the input config list have a mix of different `type`. I found only 2 places this function was use, both in the NewAuthorizationPolicies, and it seems to me the the list are of the same type. If that the case, it's probably better remove this function and call the addServiceRole(s) and addServiceRoleBinding(s) directly.

Reply:
Yeah, you're right, it's a bit over-designed here.